src/analysis/syntax_doc.ml

```diff
@@ -0,0 +1,128 @@
+        }
+      | [_, Structure _] -> 
+        Some { name = "Documentation comments"; 
+                description = "Automatically converted during parsing into attributes to allow tools to process them as documentation";
+                documentation = "https://v2.ocaml.org/releases/5.1/htmlman/doccomments.html";
+        } 
+      | (_, Module_expr _) :: (_, Module_type {mty_desc = Tmty_typeof _; _}) :: _ ->
+        Some { name = "Recovering the type of a module"; 
+                description = "The construction `module type of module-expr` expands to the module type (signature or functor type) inferred \
+                               for the module expression `module-expr`. ";
+                documentation = "https://v2.ocaml.org/releases/5.1/htmlman/moduletypeof.html";
+        }
+      | node -> 
+        let n = List.rev node in 
+        match n with 
+        | _ :: (_, Structure_item ({str_desc = Tstr_recmodule _; _}, _)) :: _  -> 
+          Some { name = "Recursive modules"; 
+                  description = "Recursive module definitions, introduced by the `module rec …and …` construction, \
+                                generalize regular module definitions `module module-name = module-expr` and module \
+                                specifications `module module-name : module-type` by allowing the defining `module-expr` \
+                                and the `module-type` to refer recursively to the module identifiers being defined.";
+                  documentation = "https://v2.ocaml.org/releases/5.1/htmlman/recursivemodules.html";
+          }
+        | _ :: _ :: (_, Value_binding {vb_expr = {exp_extra = ([Texp_newtype' _,_,_]); _}; _}) :: _ ->
+          Some { name = "Locally Abstract Type"; 
+                  description = "The expression `fun ( type typeconstr-name ) -> expr` introduces a type constructor named `typeconstr-name` which \
+                                  is considered abstract in the scope of the sub-expression, but then replaced by a fresh type variable.";
+                  documentation = "https://v2.ocaml.org/releases/5.1/htmlman/locallyabstract.html";
+          }
+        | _ :: _ :: (_, Value_binding {vb_expr = {exp_desc = (Texp_pack _); exp_extra = [(Texp_constraint {ctyp_desc = Ttyp_package _; _}, _, _)]; _}; _}) :: _ ->
+          Some { name = "First class modules"; 
+                  description = "The expression ( `module module-expr : package-type` ) converts the module (structure or functor) denoted by module \
+                                 expression `module-expr` to a value of the core language that encapsulates this module.";
+                  documentation = "https://v2.ocaml.org/releases/5.1/htmlman/firstclassmodules.html";
+          }
+      | _ -> None      
+    end
+  in
+  match info with
+  | Some info -> `Found (Printf.sprintf "%s: \n%s \n%s" info.name info.description info.documentation)
+  | None -> `No_documentation
```

Comment:
For a good separation between what's done in the "backend" (i.e. here) and what's done in the "frontend" (i.e. in `frontend/`), it could be good to return the `info` value as is here. That type can then be managed further by `Query_commands`/`Query_protocol` and be turned into a textual structure (i.e. into a JSON or a string) in `ocamlmerlin/` and in `ocaml-lsp-server`.

Btw, similarly to the other commands, the `ocamlmerlin` binary could turn this into a JSON that reflects your `info` type (`{"name" : (...), "description" : (...), "documentation" : (...)}`). That's easier to process by the plug-ins than this long string here. You can have a look at the tests of other commands such as `type-enclosing` or `locate` to see how the response value is usually structured.

@voodoos , is that also something you had in mind? Haven't you mentioned it yet, because it's quite low-level (i.e. it's a comment mostly on the implementation and only slightly on the behavior)? @PizieDust , let me know if you'd like to focus on polishing the behavior for now and think about the implementation structure later.

Reply:
@pitag-ha 
> Are you planning to add an mli-file for syntax_doc.ml once you won't change that file that much anymore?

Yes, thank you. I just pushed a commit with the `.mli` file.

Reply:
We definitely want to keep the structured information as long as possible that is as a return type for the command and with a json output that reflect it.

Reply:
> We definitely want to keep the structured information as long as possible that is as a return type for the command and with a json output that reflect it.

:+1: 

@PizieDust, let us know if you have questions about this :)

Reply:
@pitag-ha @voodoos 
This was a great idea. @jfeser left a comment [the comment](https://github.com/ocaml/ocaml-lsp/pull/1218#issuecomment-1852265222) about using [Go to Documentation]() type links instead of adding the full `url to the UX and returning the results as a `JSON will make this a breeze.

I've made some modifications and the new output should correctly return a `JSON:
```
 // Variuant types
   $ $MERLIN single syntax-document -position 1:1 \
   > -filename ./variant-types.ml < ./variant-types.ml
+  {
+    "class": "return",
+    "value": {
+      "name": "Extensible variant types",
+      "description": "Can be extended with new variant constructors using +=.",
+      "url": "https://v2.ocaml.org/manual/extensiblevariants.html"
+    },
+    "notifications": []
+  }
```
This update will be included in the next commit